[Product AI v2] Support editing the text fields

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #11996
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
This PR adds logic to support:
- Editing the text fields (name, summary and description)
- Logic to show an "Undo" button for the edited fields, and to handle its action.
- UI updates to support the above changes.

### Testing information
1. Make sure you are signed in using a Jetpack site.
2. Open the app, and start AI product creation.
3. Enter some keywords, then proceed.
4. On the preview screen, make some changes to the text fields.
5. Confirm the Undo button is shown when there are some changes.
6. Click on the Undo button.
7. Confirm it resets the content.
8. Make some changes to the name field.
9. Click on "Next" to select a different Option.
10. Confirm the "Undo" button gets hidden, and the name was updated with the generated value.
11. Go back to the previous option.
12. Confirm the text of the name still contains the edited value before switching the option.

### Images/gif
<img src="https://github.com/user-attachments/assets/e03c857d-353b-42b0-8966-5a189845dcc2" width=400 />


- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Update ViewModel to support changing fields manually

hichamboushaba committed:
Handle editing the product fields and showing the Undo option

hichamboushaba committed:
Persist the edited fields when switching options

hichamboushaba committed:
Clear text fields focus when changing the selected variant

dangermattic commented:
<!--
  0 Errors
  2 Warnings: Class <code>TextFieldState</co...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>TextFieldState</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/226">19.6</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ai/preview/AiProductPreviewViewModel.kt:
```diff
@@ -119,31 +145,75 @@ class AiProductPreviewViewModel @Inject constructor(
+        userEditedShortDescription.value = shortDescription
+    }
+
+    private fun saveUserEditedFields() {
+        generatedProduct.update { result ->
+            result?.map { product ->
+                product.copy(
+                    names = product.names.toMutableList().apply {
+                        set(selectedVariant.value, userEditedName.value ?: get(selectedVariant.value))
+                    },
+                    descriptions = product.descriptions.toMutableList().apply {
+                        set(selectedVariant.value, userEditedDescription.value ?: get(selectedVariant.value))
+                    },
+                    shortDescriptions = product.shortDescriptions.toMutableList().apply {
+                        set(selectedVariant.value, userEditedShortDescription.value ?: get(selectedVariant.value))
+                    }
+                )
+            }
+        }
+        userEditedName.value = null
+        userEditedDescription.value = null
+        userEditedShortDescription.value = null
+    }
```
I did this to follow what iOS implemented, but I also started a discussion about it internally p1721058788736309-slack-C03L1NF1EA3 to understand the rationale behind this decision.

hichamboushaba commented:
I'm converting this to draft to fix the logic in order to align with `iOS` p1721099367179549/1721058788.736309-slack-C03L1NF1EA3

hichamboushaba committed:
Support persisting the edits for the three variants

JorgeMucientes reviewed (approved):
Great job @hichamboushaba everything works as described. I do have a couple minor suggestions it might be worth adding. But nothing blocking. 

https://github.com/user-attachments/assets/fff35a4e-b242-483b-9656-f0a06423b8e0

- The undo action should have a clickable/ripple effect
- It would be nice to have the whole row clickable rather than only the "undo" text section
- Is there a reason for the Undo text to be upper case? From the Figma designs its not the case. I think we should replicate the style of the "Read text from product photo" button on the previous screen. 
- When manually undoing the changes by deleting the typed content, if the text eventually has the exact same value than the suggested one, the undo button should be hidden. 

Let me know what you think and feel free to apply any of those :)

hichamboushaba committed:
Disable allCaps for the Undo button

hichamboushaba committed:
Hide undo button when text changes to the same value as generated one

hichamboushaba commented:
Thanks @JorgeMucientes for the nice feedback, I applied most of it:

> The undo action should have a clickable/ripple effect

It already has a ripple provided by the Material `TextButton`, I think it's not noticeable just because we hide the button quickly after, check the following screenshot:
<img src=https://github.com/user-attachments/assets/766c68a9-60ef-4936-b6a6-bb05680d7426 width=360 />

> It would be nice to have the whole row clickable rather than only the "undo" text section

I'm not sure this would be a good idea, the action is destructible with no ability to rollback its changes, so making the whole area clickable could lead to accidental taps. And also in Figma, the button itself doesn't take the whole area, which makes me think this was also Joe's intention, WDYT?

> Is there a reason for the Undo text to be upper case? From the Figma designs its not the case. I think we should replicate the style of the "Read text from product photo" button on the previous screen.

Nice catch, fixed in bda2f62e8b89d197b8ad1e5890059a4f05e66bed

> When manually undoing the changes by deleting the typed content, if the text eventually has the exact same value than the suggested one, the undo button should be hidden.

Good idea, implemented in 779bc27153a8b808f3a54f2907c2007ed7014f7d

JorgeMucientes commented:
> I think it's not noticeable just because we hide the button quickly after, 

True! Completely missed it sorry for the noise. 

> I'm not sure this would be a good idea, the action is destructible with no ability to rollback its changes, so making the whole area clickable could lead to accidental taps. And also in Figma, the button itself doesn't take the whole area, which makes me think this was also Joe's intention, WDYT?

Fair point. Let's keep it as you implemented it. 

Ready to ship! :shipit:
